includes/cli/class-amp-cli-validation-command.php

```diff
@@ -222,15 +139,26 @@ public function run( $args, $assoc_args ) {
 			sprintf( 'Validating %d URLs...', $number_urls_to_crawl ),
 			$number_urls_to_crawl
 		);
-		$this->crawl_site();
+
+		$result = $url_validation_provider->with_lock(
+			function () {
+				$this->validate_urls();
+			}
+		);
+
+		if ( is_wp_error( $result ) ) {
+			WP_CLI::error( 'The site cannot be crawled at this time because validation is running in another process.' );
+			return;
+		}
```

Comment:
Because the `validate_urls` method ultimately leads to `AMP_Validation_Manager::validate_url`, which is an expensive function using `wp_remote_get`. In theory, without the lock, if someone runs the CLI command (or even multiple instances of the command) there could be multiple processes running a string of `wp_remote_get` calls at the same time. I suspect some hosting providers wouldn't like that. Do you think the lock is overcautious? Removing it would certainly simplify things.

Reply:
Forgot to note here, the lock is looking ahead to having the cron task in place, as well as any URL validation taking place within the plugins settings screen.

Reply:
OK, I'm still not sure on the need for locking but we can revisit that later. It may be preferable for any new invocation of the validation provider to abort any existing validation tasks. Then you wouldn't have the experience of validation being performed in cron, and then being blocked when attempting to initiate validation from the admin screen. 

Nevertheless, validation happening concurrently generally should not have negative consequences other than additional server load. Even if the same URL is being validated at the exact same time, what would go wrong?

Reply:
I'll be following up shortly with the cron PR that uses the lock more. That will be a good time to revisit. My fundamental assumption is that hosting platforms like VIP might see multiple background processes running `wp_remote_get` and not liking that too much, but I might be overcomplicating things because I'm sure this is not the only plugin that does things like that.